src/main/java/org/commcare/formplayer/services/MenuSessionRunnerService.java

```diff
@@ -557,20 +559,35 @@ public BaseResponseBean resolveFormGetNext(MenuSession menuSession) throws Excep
     private boolean executeAndRebuildSession(MenuSession menuSession)
             throws CommCareSessionException, RemoteInstanceFetcher.RemoteInstanceException {
         menuSession.getSessionWrapper().syncState();
-        if (menuSession.getSessionWrapper().finishExecuteAndPop(
-                menuSession.getSessionWrapper().getEvaluationContext())) {
+        StackObserver observer = new StackObserver();
+        EvaluationContext ec = menuSession.getSessionWrapper().getEvaluationContext();
+        boolean continueSession = menuSession.getSessionWrapper().finishExecuteAndPop(ec, observer);
+        clearVolatiles(menuSession.getSessionWrapper(), observer);
+        if (continueSession) {
             String smartLinkRedirect = menuSession.getSessionWrapper().getSmartLinkRedirect();
             if (smartLinkRedirect != null) {
                 menuSession.setSmartLinkRedirect(smartLinkRedirect);
             } else {
-                menuSession.getSessionWrapper().clearVolatiles();
                 menuSessionFactory.rebuildSessionFromFrame(menuSession, caseSearchHelper);
             }
             return true;
         }
         return false;
     }
 
+    private void clearVolatiles(SessionWrapper session, StackObserver observer) {
+        session.clearVolatiles();
+        for (StackFrameStep abandonedStep : observer.getRemovedSteps()) {
+            for (ExternalDataInstanceSource source : abandonedStep.getDataInstanceSources().values()) {
+                try {
+                    caseSearchHelper.clearCacheForInstanceSource(source);
+                } catch (InvalidStructureException e) {
+                    log.warn("Error clearing virtual instance cache");
```

Comment:
Should this say remote not virtual? PR description says this only clears remote instances.